File: 784-insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class pair{
    TreeNode parent;
    boolean side;
    pair(){}
    pair(TreeNode parent,boolean side){
        this.parent = parent;
        this.side=side;
    }
    
}
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        if(root==null) return new TreeNode(val);
        boolean isLeft = false;
        pair p=solve(root,val);
    
        TreeNode parent = p.parent;
        isLeft=p.side;
        TreeNode child = new TreeNode(val);
        if(isLeft){
           parent.left=child;

        }else parent.right = child;

        return root;
    }

    public pair solve(TreeNode root,int val ){
        if(root==null) return null;

        if(root.val<val && root.right==null) return new pair(root,false);
        if(root.val>val && root.left==null) return new pair(root,true);

        pair right=new pair();
        pair left=new pair();
        if(root.val<val){
            right=solve(root.right,val);
        }
        else left=solve(root.left,val);

        return left.parent==null?right:left;
    }
}

File: 784-insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if(!root) return new TreeNode(val);
        pair<TreeNode* , bool>p =solve(root,val);
        TreeNode* parent = p.first;
        TreeNode* child = new TreeNode(val);
        if(p.second){
            parent->left=child;
        }else parent->right = child;

        return root;
    }
    pair<TreeNode*,bool>solve(TreeNode* root,int val){
        if(!root) return {};

        if(root->val>val && root->left==nullptr) return {root,true};

        if(root->val<val&& root->right==nullptr) return {root,false};

        if(root->val<val) return solve(root->right,val);
        else return solve(root->left,val);
    }
};